source/Nevermore.IntegrationTests/Advanced/UseExistingTransactionFixture.cs

```diff
@@ -0,0 +1,167 @@
+        {
+            ++CommitTransactionCalled;
+            await Transaction.CommitAsync(cancellationToken);
+        }
+
+        public void Dispose()
+        {
+            ++DisposeCalled;
+            Transaction.Dispose();
+            Connection.Dispose();
+        }
+    }
+
+    [Test]
+    [TestCase(false)]
+    [TestCase(true)]
+    public async Task OpeningReadTransaction_FromExistingTransaction_Throws(bool takeOwnershipOfExistingConnectionAndTransaction)
+    {
+        await using var sqlConnection = new SqlConnection(ConnectionString);
+        await sqlConnection.OpenAsync(cancellationToken);
+        await using var sqlTransaction = sqlConnection.BeginTransaction();
+
+        var readTransaction = (ReadTransaction) Store.CreateReadTransactionFromExistingConnectionAndTransaction(
+            new TestConnectionAndTransaction(sqlConnection, sqlTransaction),
+            takeOwnershipOfExistingConnectionAndTransaction);
+
+        Assert.Multiple(() =>
+        {
+            Assert.ThrowsAsync<InvalidOperationException>(async () => await readTransaction.OpenAsync(cancellationToken));
+            Assert.Throws<InvalidOperationException>(() => readTransaction.Open());
+        });
+    }
+
+    [Test]
+    [TestCase(false)]
+    [TestCase(true)]
+    public async Task OpeningWriteTransaction_FromExistingTransaction_Throws(bool takeOwnershipOfExistingConnectionAndTransaction)
+    {
+        await using var sqlConnection = new SqlConnection(ConnectionString);
+        await sqlConnection.OpenAsync(cancellationToken);
+        await using var sqlTransaction = sqlConnection.BeginTransaction();
+
+        var writeTransaction = (WriteTransaction) Store.CreateWriteTransactionFromExistingConnectionAndTransaction(
+            new TestConnectionAndTransaction(sqlConnection, sqlTransaction),
+            takeOwnershipOfExistingConnectionAndTransaction);
+
+        Assert.Multiple(() =>
+        {
+            Assert.ThrowsAsync<InvalidOperationException>(async () => await writeTransaction.OpenAsync(cancellationToken));
+            Assert.Throws<InvalidOperationException>(() => writeTransaction.Open());
+        });
+    }
+
+    [Test]
+    public async Task CommittingNonOwnedTransaction_Throws()
+    {
+        var sqlConnection = new SqlConnection(ConnectionString);
+        await sqlConnection.OpenAsync(cancellationToken);
+        var sqlTransaction = sqlConnection.BeginTransaction();
+        using var connectionAndTransaction = new TestConnectionAndTransaction(sqlConnection, sqlTransaction);
+
+        var nonOwnedTransaction = Store.CreateWriteTransactionFromExistingConnectionAndTransaction(
+            connectionAndTransaction, false);
+
+        Assert.Multiple(() =>
+        {
+            Assert.ThrowsAsync<InvalidOperationException>(async () => await nonOwnedTransaction.CommitAsync(cancellationToken));
+            Assert.Throws<InvalidOperationException>(() => nonOwnedTransaction.Commit());
```

Comment:
InvalidOperationException can happen for lots of reasons. As above suggest FluentAssertions, then you could do `.WithMessage()` and assert that you got the right one.

Reply:
I know some people don't like asserting on exception messages but I'm ok with it, I don't envision these changing and causing friction often if ever, so I'll add this 👍

Reply:
To be clear: Asserting on exception messages is not good, it's brittle and annoying because messages change over time, god forbid we ever translate octopus to a non-english language.
But asserting on something generic like `InvalidOperationException` is also not good because there can be many different things that throw it, and you don't know if you've caught the right one.

My personal preference is either:
- Throw a specific exception that you can catch without ambiguity
- Throw an exception which carries some sort of programmatic error indicator, like a numeric error code, and assert on that. COMException does this with HRESULTs and I'm pretty sure the SQL/Database exceptions carry numeric error codes for similar reasons.

That said, pick your poison and do whatever is most practical here. If you're confident that transaction commit won't throw `InvalidOperationException` from any paths other than the one you expect, that's good enough.